src/main/java/org/embulk/filter/column/JsonColumn.java

```diff
@@ -39,33 +44,46 @@ public JsonColumn(String path, Type type, Value defaultValue)
 
     public JsonColumn(String path, Type type, Value defaultValue, String src)
     {
-        this.path = path;
+        CompiledPath compiledPath = PathCompiler.compile(path);
+        CompiledPath compiledSrc = src == null ? compiledPath : PathCompiler.compile(src);
+        this.path = compiledPath.toString();
         this.type = type;
         this.defaultValue = (defaultValue == null ? ValueFactory.newNil() : defaultValue);
-        this.src = (src == null ? path : src);
+        this.src = compiledSrc.toString();
 
         this.pathValue = ValueFactory.newString(path);
-        this.parentPath = parentPath(path);
-        this.baseName = baseName(path);
+        this.parentPath = compiledPath.getParentPath();
+        this.baseName = compiledPath.getTailPath();
         if (this.baseName.equals("[*]")) {
             throw new ConfigException(String.format("%s wrongly ends with [*], perhaps you can remove the [*]", path));
         }
-        this.baseIndex = baseIndex(path);
+        this.baseIndex = compiledPath.baseIndex();
         this.parentPathValue = ValueFactory.newString(parentPath);
+        String baseName = getPropertyOrBaseName(compiledPath, this.baseName);
         this.baseNameValue = ValueFactory.newString(baseName);
 
         this.srcValue = ValueFactory.newString(this.src);
-        this.srcParentPath = parentPath(this.src);
-        this.srcBaseName = baseName(this.src);
-        this.srcBaseIndex = baseIndex(this.src);
+        this.srcParentPath = compiledSrc.getParentPath();
+        this.srcBaseName = compiledSrc.getTailPath();
+        this.srcBaseIndex = compiledSrc.baseIndex();
         this.srcParentPathValue = ValueFactory.newString(this.srcParentPath);
-        this.srcBaseNameValue = ValueFactory.newString(this.srcBaseName);
+        String srcBaseName = getPropertyOrBaseName(compiledSrc, this.srcBaseName);
+        this.srcBaseNameValue = ValueFactory.newString(srcBaseName);
 
         if (! srcParentPath.equals(parentPath)) {
             throw new ConfigException(String.format("The branch (parent path) of src \"%s\" must be same with of name \"%s\" yet", src, path));
         }
     }
 
+    // Fixme baseNameValue would become like a "[0]" if baseName equals [0] (ArrayPropertyToken)
```

Comment:
baseNameValue and srcBaseNameValue has been used as the key of the map in visitMap of JsonVisitor. Therefore, String for the Array (eg. "[0]") will basically not be used. Since the original implementation had this problem, I could not decide how to do.

https://github.com/kysnm/embulk-filter-column/blob/fbd5e3b5bc37bc3126c57bdaed8c76eddb9c9ce0/src/main/java/org/embulk/filter/column/JsonVisitor.java#L276-L333